### PR TITLE
feat(DeviceConnector/Device): Devices, DeviceTypes & Monitoring Part 1

### PR DESCRIPTION
**Pull request by Gozzim**
## Changes Proposed:

- Multiple type additions
- Various functions for Device and DeviceTypes
- Proper handling of not existing or unavailable devices
- DeviceConnector now connects to DeploymentServer via a Duplex stream instead of a WebSocket
- Various changes for future monitor handling & piping
- Device-Status is now updated locally and remotely
- Code-CleanUp, Refactoring and Type-Changes

## How to Test the Changes:

1. StartUp processes as per usual
2. Deploy and change Code
3. Result on the device should be the same as on older versions

## Known Issues and TODO List:

- Monitoring is not in its finished state
- Monitoring output is not yet piped to deployment-server
- Deployment-Status is not yet updated on the server accordingly

## How to Test PRs
1) First fetch the Pull-Request from the main repo, replacing `xx` with the ID of the PR:\
`git fetch git@github.com:eclipsesource/cdtcloud-deploymentserver.git pull/xx/head:pr_xx`
2) Checkout the Pull-Request, again replacing `xx` with the PR-ID:\
`git checkout pr_xx`
3) Perform the necessary tests for the PR

### CODE REVIEW

File: packages/theia-extension/yarn.lock [hunk 1]
```diff
@@ -2,3 +2,189 @@
 # yarn lockfile v1
```
Comment (WoH) [inline packages/theia-extension/yarn.lock:2]: remove

File: packages/device-connector/src/util/common.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+export enum grpcStatus {
```
Comment (WoH) [inline packages/device-connector/src/util/common.ts:1]: prefer capitalized enums

File: packages/device-connector/src/deployment-server/service.ts [hunk 3]
```diff
@@ -81,16 +93,14 @@ export const fetchDeviceType = async (typeId: string): Promise<any> => {
...
   })
 
-  return await resp.json()
+  return await resp.json() as DeviceType
 }
 
-export const deleteDeviceRequest = async (deviceId: string): Promise<any> => {
+export const deleteDeviceRequest = async (deviceId: string): Promise<void> => {
   const address = env.SERVER_URI != null ? env.SERVER_URI : '127.0.0.1:3001'
   const url = `http://${address}/devices/${deviceId}`
```
Comment (WoH) [inline packages/device-connector/src/deployment-server/service.ts:101]: We may want to be able configure another protocol (HTTPS)

File: packages/device-connector/src/deployment-server/service.ts [hunk 4]
```diff
@@ -67,10 +79,10 @@ export const fetchAllDeviceTypes = async (): Promise<any> => {
...
   })
 
-  return await resp.json()
+  return await resp.json() as DeviceType[]
 }
 
-export const fetchDeviceType = async (typeId: string): Promise<any> => {
+export const fetchDeviceType = async (typeId: string): Promise<DeviceType> => {
   const address = env.SERVER_URI != null ? env.SERVER_URI : '127.0.0.1:3001'
   const url = `http://${address}/device-types/${typeId}`
```
Comment (WoH) [inline packages/device-connector/src/deployment-server/service.ts:87]: protocol

File: packages/device-connector/src/deployment-server/service.ts [hunk 5]
```diff
@@ -31,32 +45,30 @@ export const sendNewDeviceRequest = async (typeId: string): Promise<any> => {
...
+    body: JSON.stringify(data)
   })
 
-  return await resp.json()
+  return await resp.json() as DeviceResponse
 }
 
-export const fetchAllDeviceTypes = async (): Promise<any> => {
+export const fetchAllDeviceTypes = async (): Promise<DeviceType[]> => {
   const address = env.SERVER_URI != null ? env.SERVER_URI : '127.0.0.1:3001'
```
Comment (WoH) [inline packages/device-connector/src/deployment-server/service.ts:72]: protocol

File: packages/device-connector/src/deployment-server/service.ts [hunk 6]
```diff
@@ -31,32 +45,30 @@ export const sendNewDeviceRequest = async (typeId: string): Promise<any> => {
...
 
-  return await resp.json()
+  return await resp.json() as DeviceResponse
 }
 
-export const setDeviceRequest = async (typeId: string): Promise<any> => {
+export const setDeviceRequest = async (deviceId: string, data: object): Promise<DeviceResponse> => {
   const address = env.SERVER_URI != null ? env.SERVER_URI : '127.0.0.1:3001'
-  const url = `http://${address}/devices/${typeId}`
+  const url = `http://${address}/devices/${deviceId}`
```
Comment (WoH) [inline packages/device-connector/src/deployment-server/service.ts:58]: protocol

File: packages/device-connector/src/deployment-server/service.ts [hunk 7]
```diff
@@ -17,10 +31,10 @@ export const sendNewDeviceTypeRequest = async (fqbn: string, name: string): Prom
...
   })
 
-  return await resp.json()
+  return await resp.json() as DeviceType
 }
 
-export const sendNewDeviceRequest = async (typeId: string): Promise<any> => {
+export const sendNewDeviceRequest = async (typeId: string): Promise<DeviceResponse> => {
   const address = env.SERVER_URI != null ? env.SERVER_URI : '127.0.0.1:3001'
   const url = `http://${address}/devices`
```
Comment (WoH) [inline packages/device-connector/src/deployment-server/service.ts:39]: protocol

**Comment by WoH** [comment]
Additionally, as discussed:
- [ ] `#prop` over `_prop`
- [ ] Share types

File: packages/device-connector/src/deployment-server/service.ts (hunk 3 above)
Comment (Gozzim) [reply packages/device-connector/src/deployment-server/service.ts:101]: Yeah, I am aware of this and it is on my ToDo list for changes in the near future (see Trello)

**Comment by Gozzim** [comment]
> Additionally, as discussed:
> 
> * [ ]  `#prop` over `_prop`
> * [ ]  Share types

Yes, Part 2 of the PR will have massive changes in those cases anyway. And it will be included there

File: packages/theia-extension/yarn.lock (hunk 1 above)
Comment (Gozzim) [reply packages/theia-extension/yarn.lock]: resolved in 03c659a4b443a8c48d6aeaddcf6f30e1de3a948d

File: packages/device-connector/src/util/common.ts (hunk 2 above)
Comment (Gozzim) [reply packages/device-connector/src/util/common.ts]: resolved in 03c659a4b443a8c48d6aeaddcf6f30e1de3a948d

**Review by WoH — APPROVED** [review]

**Pull request merged**